### ISSUE TITLE
Satisfying pattern for WPS412 and WPS203?

### ISSUE THREAD

**Issue by cbows** [wemake-services/wemake-python-styleguide#1478]
# Question

This isn't really a bug report, more a question.

Assuming the following file structure:
```
package/
  |- module.py
  |- subpackage/
    |- __init__.py
    |- foo.py
    |- bar.py
    |- baz.py
    |- ...
```

Subpackage contains a bunch of modules `foo`, `bar`, `baz`, etc. and `module` wants to import them, so I could do
```python
from package.subpackage import foo
from package.subpackage import bar
from package.subpackage import baz

# now I can call
foo.Foo()
bar.Bar()
baz.Baz()
```
but this pattern will quickly raise WPS203 (too many imports) on large projects where I might also have other subpackages. To circumvent this, I am tempted to do the following
```python
from package import subpackage

subpackage.foo.Foo()
subpackage.bar.Bar()
subpackage.baz.Baz()
```
This, however, requires logic in `__init__.py`, e.g.
```python
from package.subpackage import foo, bar, baz
```
which violates WPS412.

I see this problem in many APIs (e.g. `scikit-learn`) and they all solve it more or less by including logic in `__init__.py`. Nonetheless, I do appreciate having no logic in `__init__.py` because I have struggled with various problems it can cause myself, so I worked around the issue by introducing the following structure
```
package/
  |- module.py
  |- subpackage.py
  |- _subpackage/
    |- __init__.py
    |- foo.py
    |- bar.py
    |- baz.py
```
and putting the import logic into `subpackage.py`
```python
from _subpackage import foo, bar, baz
```
It works, but also introduces a rather unintuitive project structure and some duplication.

What is the recommended pattern here or would you consider this good enough? Would an exception to allow something like `from . import *` in `__init__.py` make sense?

**Comment by sobolevn** [comment]
@cbows thanks a lot for raising this question! 

I would say that if this happens in a single file, then you can just ignore this violation and that's it.
It might happen in cases like: dependency injection, component registration, complex business logic with lots of steps and actors, etc.

But, if this happens a lot in your project, it means that you have an architecture problem. The whole idea of this rule is too decrease coupling of modules inside your app. If it is high, you would have many imports. And you have to keep it low.

That's why fixing it by creating `__init__.py` with reimports is not a good idea. You have to solve the coupling problem. It is the root cause of multiple imports.

**Comment by cbows** [comment]
Thank you. I think I understand the reasoning behind theses rules better now.

I agree about the tight coupling problem you mentioned. But I just realized that in my case it seems to be more of a UI problem I want to solve. For example I have subpackages `models` and `schemas`. One is for database models and the other for schema definitions of objects passed to clients in a web app. There is certain similarity in those, e.g. both contain a `users` module with models and schemas for users. They are often used together so have to be imported together (hence double amount of imports) and I still want them to be easily distinguishable (I would like to be able to do `schemas.users.User` and `models.users.User`). I feel comfortable to add exceptions manually here or even use my workaround.
